x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -992,35 +977,26 @@ private void findTokenFromRefreshToken(String refreshToken, Iterator<TimeValue>
-                findTokenFromRefreshToken(refreshToken, securityTokensIndex, backoff, listener);
+            logger.debug("Assuming a refresh token [{}] provided from a client", refreshToken);
+            final TransportVersion refreshTokenVersion;
+            final String unencodedRefreshToken;
+            final Tuple<TransportVersion, String> versionAndRefreshTokenTuple;
+            try {
+                versionAndRefreshTokenTuple = unpackVersionAndPayload(refreshToken);
+                refreshTokenVersion = versionAndRefreshTokenTuple.v1();
+                unencodedRefreshToken = versionAndRefreshTokenTuple.v2();
+            } catch (IOException e) {
+                logger.debug(() -> "Could not decode refresh token [" + refreshToken + "].", e);
+                listener.onResponse(SearchHits.EMPTY_WITH_TOTAL_HITS);
+                return;
+            }
+            if (refreshTokenVersion.before(VERSION_TOKENS_INDEX_INTRODUCED) || unencodedRefreshToken.length() != TOKEN_LENGTH) {
+                logger.debug("Decoded refresh token [{}] with version [{}] is invalid.", unencodedRefreshToken, refreshTokenVersion);
+                listener.onResponse(SearchHits.EMPTY_WITH_TOTAL_HITS);
             } else {
-                logger.debug("Assuming a refresh token [{}] provided from a client", refreshToken);
-                final TransportVersion refreshTokenVersion;
-                final String unencodedRefreshToken;
-                final Tuple<TransportVersion, String> versionAndRefreshTokenTuple;
-                try {
-                    versionAndRefreshTokenTuple = unpackVersionAndPayload(refreshToken);
-                    refreshTokenVersion = versionAndRefreshTokenTuple.v1();
-                    unencodedRefreshToken = versionAndRefreshTokenTuple.v2();
-                } catch (IOException e) {
-                    logger.debug(() -> "Could not decode refresh token [" + refreshToken + "].", e);
-                    listener.onResponse(SearchHits.EMPTY_WITH_TOTAL_HITS);
-                    return;
-                }
-                if (refreshTokenVersion.before(VERSION_TOKENS_INDEX_INTRODUCED) || unencodedRefreshToken.length() != TOKEN_LENGTH) {
-                    logger.debug("Decoded refresh token [{}] with version [{}] is invalid.", unencodedRefreshToken, refreshTokenVersion);
-                    listener.onResponse(SearchHits.EMPTY_WITH_TOTAL_HITS);
-                } else {
-                    // TODO Remove this conditional after backporting to 7.x
-                    if (refreshTokenVersion.onOrAfter(VERSION_HASHED_TOKENS)) {
-                        final String hashedRefreshToken = hashTokenString(unencodedRefreshToken);
-                        findTokenFromRefreshToken(hashedRefreshToken, securityTokensIndex, backoff, listener);
-                    } else {
-                        findTokenFromRefreshToken(unencodedRefreshToken, securityTokensIndex, backoff, listener);
-                    }
```

Comment:
As an "extra", I've also removed this if-condition branch as it was not viable (commit https://github.com/elastic/elasticsearch/pull/97422/commits/5e7f77a9af29cf87ebbef5d6378d6d0d8de85815): refresh tokens prepended with a version are always hashed (never raw).

Reply:
Either way that if-block (as it existed before) doesn't actually make sense to me:

`VERSION_TOKENS_INDEX_INTRODUCED` = `VERSION_HASHED_TOKENS` = `7.2`, so `refreshTokenVersion.onOrAfter(VERSION_HASHED_TOKENS)` was always guaranteed to be true since otherwise, we would have not gotten to the else clause of, i.e.,:

```
if (refreshTokenVersion.before(7.2) || ...) {

} else { 
  // refreshTokenVersion.onOrAfter(7.2) must hold regardless of whether we backport anything or not
} 
```

tl;dr: thanks for removing this 👍

Reply:
Yes, the else branch was not possible, as you describe, there are multiple version constants equaling `7.2`, introduced in independent work streams for token format changes that got merged in the same release, and so it's not possible that one be false and the other true.

> This is pertaining to the "// TODO Remove this conditional after backporting to 7.x" and the functionality that was introduced by https://github.com/elastic/elasticsearch/pull/54532 and backported to 7.x with https://github.com/elastic/elasticsearch/commit/ac87c100396992a1d96954739be7e8569b40bff5, correct?

I assumed as much myself, I don't recall anything about the context of the comment...